### ISSUE TITLE
HeapProfiler.takeHeapSnapshot usage?

### ISSUE THREAD

**Issue by alessioalex** [paulirish/automated-chrome-profiling#3]
Hei @paulirish I've tried to take a heap snapshot and save it to disk, but with no success. I didn't find much documentation on this.

Can you give me a hand please? Thanks a bunch!

**Comment by cyrus-and** [comment]
@alessioalex maybe you can add some details. What did you try so far? What is the problem exactly?

**Comment by vladikoff** [comment]
https://developer.chrome.com/devtools/docs/heap-profiling#basics

**Comment by alessioalex** [comment]
@vladikoff I think you misunderstood my problem. I know how to take a heapdump with Chrome DevTools / Node, I just don't know how to make it an automated process by using the Chrome Remote Debugging protocol.

I saw that there was a method `HeapProfiler.takeHeapSnapshot` when trying out the `chrome-remote-interface` NPM module (repl cli version), I just couldn't get it to work. I also didn't find much documentation about using `chrome-remote-interface` around the web apart from this repo.

TL;DR I want a small snippet with `chrome-remote-interface` showing how to take a heapdump and save it to disk. Basically what this repo does for CPU profiling, but for memory usage (heap dump).

@paulirish @vladikoff @cyrus-and can you guys help me out? Thanks a bunch

**Comment by cyrus-and** [comment]
@alessioalex in the meantime, the documentation of `chrome-remote-interface` can be found [here](https://github.com/cyrus-and/chrome-remote-interface).

**Comment by dchristian-ias** [comment]
@alessioalex Just curious, did you ever find a solution to this or find a way to successfully get the contents of the `takeHeapSnapshot` using the `chrome-remote-interface`? I am having the same issues as you at the moment..

**Comment by cyrus-and** [comment]
@dwcIntegral have you read the documentation of the [HeapProfiler](https://chromedevtools.github.io/debugger-protocol-viewer/tot/HeapProfiler/) module? It looks like the content of the heap snapshot is delivered via events, maybe progressively via [addHeapSnapshotChunk](https://chromedevtools.github.io/debugger-protocol-viewer/tot/HeapProfiler/#event-addHeapSnapshotChunk) (cannot try right now). Just make sure to [enable](https://chromedevtools.github.io/debugger-protocol-viewer/tot/HeapProfiler/#method-enable) the events first.

**Comment by alessioalex** [comment]
@dwcIntegral I think I did, here's the snippet I found in one of my repos: https://gist.github.com/alessioalex/acded39027c0fc853ac2244c40f6438c

**Comment by cowchimp** [comment]
This lib I created can be used to take a JS heap snapshot from code. Internally it uses the DevTools protocol + the DevTools front-end to parse the raw snapshot data.
https://github.com/cowchimp/headless-devtools

**Comment by vbsteven** [comment]
I figured out how to do this today and this is what I came up with. 
```
const CDP = require('chrome-remote-interface');
const fs = require('fs');

const options = {
        host: '127.0.0.1',
        port: 9229
}

CDP(options, async (client) => {
    const {HeapProfiler} = client;
    try {
        console.log("Connected");
        HeapProfiler.addHeapSnapshotChunk((c) => {
                console.log("got chunk");
                fs.appendFileSync("/tmp/heap.txt", c.chunk);
        });
        HeapProfiler.reportHeapSnapshotProgress();
        const result = await HeapProfiler.takeHeapSnapshot();
        console.log(result);
    } catch (err) {
        console.error(err);
    } finally {
        await client.close();
    }
}).on('error', (err) => {
    console.error(err);
});
```

**Comment by alessioalex** [comment]
hey @vbsteven I tried you example but I only get 'Connected' as a message. Any tips? thanks

**Comment by vbsteven** [comment]
@alessioalex have you tried playing around in the chrome-remote-interface repl? That's how I figured out how to take a snapshot for our situation. 

The script that I posted is what I ended up with after some playing around in the repl. The tricky part was to call `reportHeapSnapshotProgress()` first otherwise the `addHeapSnapshotChunk()` callback wouldn't fire.

We have used this script successfully on Node 12 running inside a Docker container on ECS. First we enabled inspect on the process by sending it a USR1 signal and then we executed the script inside the container.

For CPU profiling we have used something similar. Does this script work for you?

```
const CDP = require('chrome-remote-interface');
const fs = require('fs');

const options = {
        host: '127.0.0.1',
        port: 9229
}

// profile duration in milliseconds
const duration = 10000;

function delay(t) {
   return new Promise(function(resolve) {
       setTimeout(function() {
           resolve();
       }, t);
   });
}

CDP(options, async (client) => {
    const {Profiler} = client;
    try {
            console.log("Connected");
            await Profiler.enable();
            await Profiler.setSamplingInterval({interval: 100});
            await Profiler.start();
            console.log("profiling started");
            await delay(duration);
            const data = await Profiler.stop();
            console.log("profiling ended");
            console.log(data);
            const filename = "/tmp/profile.cpuprofile";
            const string = JSON.stringify(data.profile, null, 2);
            fs.writeFileSync(filename, string);
            console.log(`CPU Profile written to ${filename}`);
    } catch (err) {
        console.error(err);
    } finally {
        await client.close();
    }
}).on('error', (err) => {
    console.error(err);
});
```

**Comment by cyrus-and** [comment]
@vbsteven , @alessioalex I'd use `await` with `HeapProfiler.reportHeapSnapshotProgress();` to make sure the call is processed before issuing `HeapProfiler.takeHeapSnapshot()`.
